geomesa-core/src/main/scala/org/locationtech/geomesa/core/index/FilterHelper.scala

```diff
@@ -205,6 +205,70 @@ object FilterHelper {
       }
     }
 
+  /**
+   * Finds the filter with the lowest known cost and returns the rest in the same order they were in.  If
+   * there are multiple filters with the same lowest cost then the first will be selected.  If no filters
+   * have a known cost or if ``s`` is empty then (None, s) will be returned.
+   */
+  def findBest(cost: Filter => Option[Long])(s: Seq[Filter]): CostAnalysis = {
+    if (s.isEmpty) {
+      CostAnalysis.unknown(s)
+    } else {
+      val head = s.head
+      val tail = s.tail
+
+      val headAnalysis = cost(head).map(c => new KnownCost(head, tail, c)).getOrElse(CostAnalysis.unknown(s))
+      val tailAnalysis = findBest(cost)(tail)
+
+      if (headAnalysis <= tailAnalysis) {
+        headAnalysis
+      } else {
+        // tailAnaysis must have a known cost
+        val ta = tailAnalysis.asInstanceOf[KnownCost]
+        new KnownCost(ta.best, head +: ta.otherFilters, ta.cost)
+      }
+    }
+  }
+
+  /**
+    * @param bestFilter the [[Filter]] with the lowest cost
+    * @param otherFilters all other [[Filter]]s
+    */
+  sealed abstract case class CostAnalysis(bestFilter: Option[Filter], otherFilters: Seq[Filter]) {
+
+    /**
+      * @param rhs the [[CostAnalysis]] to compare to
+      * @return ``true`` if ``this`` has a lower or the same cost as ``rhs``
+      */
+    def <=(rhs: CostAnalysis): Boolean
+
+    def extract: (Option[Filter], Seq[Filter]) = (bestFilter, otherFilters)
+  }
+
+  class KnownCost(val best: Filter, others: Seq[Filter], val cost: Long) extends CostAnalysis(Some(best), others) {
```

Comment:
I struggled with this a little.  I could make the parent class a case class or the child classes but not both and it wasn't clear to me which way was best.